docs/notebooks/tutorials/04_Omnigenity_Optimization.ipynb

```diff
@@ -0,0 +1,422 @@
+   "metadata": {
+    "pycharm": {
+     "name": "#%% md\n"
+    }
+   },
+   "source": [
+    "Now that the equilibrium is initialized, we need to solve the fixed-boundary vacuum equilibrium: "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "constraints = get_fixed_boundary_constraints(eq)\n",
+    "eq, _ = eq.solve(objective=\"vacuum\", constraints=constraints, verbose=3)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Let us make a copy of this initial equilibrium so that we can compare our final solution to it later and see how well the omnigenity optimization worked. \n",
+    "By plotting the $|B|$ contours in Boozer coordinates, we can see that this equilibrium is already somewhat close to being omnigenous with poloidal contours, but far from perfect. "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "eq0 = eq.copy()\n",
+    "plot_boozer_surface(eq0, fieldlines=4)  # defaults to the rho=1 surface"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "In order to optimize the equilibrium for omnigenity, we need to create a target omnigenous magnetic field. \n",
+    "The `OmnigenousField` class has two attributes that represent parameters in the omnigenous magnetic field model: \n",
+    "- `B_lm` specifies the shape of the \"magnetic well\" on each flux surface. \n",
+    "- `x_lmn` specifies how the well shape varies along different field lines. \n",
+    "\n",
+    "We need to specify the resolution of the parameter space and the helicity of the field. \n",
+    "The helicity is given by the tuple of integers $(M, N)$, and is set to $(0, N_{FP})$ for omnigenity with poloidal contours in this example. \n",
+    "The typical case for helical contours would be $(1, N_{FP})$, and for toroidal contours would be $(1, 0)$. \n",
+    "\n",
+    "We provide initial values for the well shape parameters `B_lm` so that we can set the mirror ratio. \n",
+    "The total number of parameters is `B_lm.size = M_well * (L_well + 1)`. \n",
+    "In this example, the well shape on each surface is represented by three spline knots, and there is $\\mathcal{O}(\\rho)$ variation across the flux surfaces. \n",
+    "Here we set only the constant terms of the Chebyshev polynomials such that the initial target field has the same magnetic well from $B_{\\mathrm{min}}=0.8\\mathrm{~T}$ to $B_{\\mathrm{max}}=1.2\\mathrm{~T}$ on each surface. \n",
+    "\n",
+    "The `x_lmn` parameters are left to their default values of 0. "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "field = OmnigenousField(\n",
+    "    L_well=1,  # radial resolution of B_lm parameters\n",
+    "    M_well=3,  # number of spline knots on each flux surface\n",
```

Comment:
I am assuming you are not explicitly specifying B_lm, just changing the well resolution. 
It would be nice to have something like 
`B_lm = np.array([[1-mirror_ratio, 1-mirror_ratio+0.05, 1, 1+mirror_ratio], [1-mirror_ratio, 1-mirror_ratio+0.1, 1, 1+mirror_ratio]])`
`eq._B_lm = B_lm`